lib/vrl/stdlib/src/for_each.rs

```diff
@@ -0,0 +1,189 @@
+    fn identifier(&self) -> &'static str {
+        "for_each"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::OBJECT | kind::ARRAY,
+                required: true,
+            },
+            Parameter {
+                keyword: "recursive",
+                kind: kind::BOOLEAN,
+                required: false,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "iterate object",
+                source: r#"count = 0; for_each({ "a": 1, "b": 2 }) -> |_key, value| { count = count + value }; count"#,
+                result: Ok("3"),
+            },
+            Example {
+                title: "recursively iterate object",
+                source: r#"count = 0; for_each({ "a": 1, "b": { "c": 2, "d": 3 } }, recursive: true) -> |_, value| { count = count + (int(value) ?? 0) }; count"#,
+                result: Ok("6"),
+            },
+            Example {
+                title: "iterate array",
+                source: r#"count = 0; for_each([1,2,3]) -> |index, value| { count = count + index + value }; count"#,
+                result: Ok("9"),
+            },
+            Example {
+                title: "recursively iterate array",
+                source: r#"count = 0; for_each([1,2,[3,4]], recursive: true) -> |index, value| { count = count + index + (int(value) ?? 0) }; count"#,
+                result: Ok("14"),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: (&mut state::LocalEnv, &mut state::ExternalEnv),
+        _ctx: &mut FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let recursive = arguments.optional("recursive");
+        let closure = arguments.required_closure()?;
+
+        Ok(Box::new(ForEachFn {
+            value,
+            closure,
+            recursive,
+        }))
+    }
+
+    fn closure(&self) -> Option<closure::Definition> {
+        use closure::{Definition, Input, Output, Variable, VariableKind};
+
+        Some(Definition {
+            inputs: vec![Input {
+                parameter_keyword: "value",
+                kind: Kind::object(Collection::any()).or_array(Collection::any()),
+                variables: vec![
+                    Variable {
+                        kind: VariableKind::TargetInnerKey,
+                    },
+                    Variable {
+                        kind: VariableKind::TargetInnerValue,
+                    },
+                ],
+                output: Output::Kind(Kind::any()),
+                example: Example {
+                    title: "iterate array",
+                    source: r#"for_each([1, 2]) -> |index, value| { .foo = to_int!(.foo) + index + value }"#,
+                    result: Ok("null"),
+                },
```

Comment:
Same question as in https://github.com/vectordotdev/vector/pull/12388/files/c367a429b690a7f1b1ce9f3eaf788ebd18d0a86e#r858515061:

Do we benefit from having another example here? Could use `self.examples()[0]` otherwise

Reply:
Yeah, they serve different purposes. Specifically, this one is explicitly tied to a given input variant, and it is also used in a diagnostic message related to invalid closures. The other one has no constraints on what it shows as an example, so it might mismatch this input variant and/or might not be fit to show in a diagnostic message.

They are the same right now, but there's no guarantee that'll stay the case.